main/src/com/google/refine/templating/Parser.java

```diff
@@ -33,15 +33,39 @@
 
 package com.google.refine.templating;
 
-import java.util.ArrayList;
-import java.util.List;
-
 import com.google.refine.expr.MetaParser;
 import com.google.refine.expr.ParsingException;
 import com.google.refine.grel.ast.FieldAccessorExpr;
 import com.google.refine.grel.ast.VariableExpr;
 
+import java.util.ArrayList;
+import java.util.List;
+
+/**
+ * Parser class
+ */
 public class Parser {
+
+    /**
+     * findEndBrace method
+     * Finds the first true ending brace in a cell, ignoring escaped end braces
+     *
+     * @param expression string that will be interrogated for first true end brace
+     * @param lowIndex   index from which the interrogation will start
+     * @return index of the first true end brace
+     * CS427 Issue Link: https://github.com/OpenRefine/OpenRefine/issues/3381
+     */
+    private static int findEndBrace(final String expression, final int lowIndex) {
+        int closeBrace = expression.indexOf('}', lowIndex);
+        if (expression.charAt(closeBrace - 1) == '\\') {
```

Comment:
What happens if no `}` character could be found? In the case of an unfinished template, we should make sure we fail graciously.

Reply:
Hey @wetneb I just added a parsing exception in case no end brace can be found. Is that what you meant by fail graciously? If not I can go back and implement whatever you were thinking.